### PR TITLE
Implement String.try_convert

### PR DESCRIPTION
**Pull request by nanobowers**
Implemented `String.try_convert`, which is structurally similar to `Array.try_convert`.
Related to #217

### CODE REVIEW

**Comment by ryangjchandler** [comment]
@nanobowers There's a small merge conflict preventing this from being merged right now, would you be able to take a look at that please?

File: src/string_object.cpp [hunk 1]
```diff
@@ -2517,4 +2517,32 @@ Value StringObject::sum(Env *env, Value val) {
...
+
+    if (conversion->is_string() || conversion->is_nil()) {
+        return conversion;
+    }
+
+    auto original_item_class_name = val->klass()->inspect_str();
+    auto new_item_class_name = conversion->klass()->inspect_str();
+    env->raise(
+        "TypeError",
+        "can't convert {} to Array ({}#to_str gives {})",
```
Comment (seven1m) [inline src/string_object.cpp:2542]: I think this should say `String` instead of `Array`.

File: src/string_object.cpp [hunk 2]
```diff
@@ -2517,4 +2517,32 @@ Value StringObject::sum(Env *env, Value val) {
...
+        original_item_class_name,
+        original_item_class_name,
+        new_item_class_name);
+}
```
Comment (seven1m) [inline src/string_object.cpp:2546]: This whole method is very similar to `Object::to_str`, I wonder if we can just call that method inside here?

Comment (nanobowers) [reply src/string_object.cpp:2546]: I didn't previously think to do that, but just tried it and it failed one of the specs:
```
Failed specs:
String.try_convert
  sends #to_str to the argument and returns the result if it's nil
    can't convert MockObject to String (MockObject#to_str gives NilClass) (TypeError)
      spec/core/string/try_convert_spec.rb:22:in `try_convert'
```
It seems that the behavior is subtly different in how nils are handled between the different approaches.

Comment (seven1m) [reply src/string_object.cpp:2542]: Ahhh, ok! Very well, let's just use your code then.

**Pull request merged**